File: src/MagnusAddon/note/vocabulary/vocabnote_kanji.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from ex_autoslot import AutoSlots
from sysutils import ex_str, kana_utils
from sysutils.collections.queryable.q_iterable import query

if TYPE_CHECKING:
    from note.vocabulary.vocabnote import VocabNote
    from sysutils.collections.queryable.collections.q_list import QList
    from sysutils.weak_ref import WeakRef

class VocabNoteKanji(AutoSlots):
    def __init__(self, vocab: WeakRef[VocabNote]) -> None:
        self.__vocab = vocab

    @property
    def _vocab(self) -> VocabNote: return self.__vocab()

    def extract_main_form_kanji(self) -> QList[str]:
        clean = ex_str.strip_html_and_bracket_markup(self._vocab.get_question())
        return query(clean).where(kana_utils.character_is_kanji).to_list() #[char for char in clean if kana_utils.character_is_kanji(char)]

    def extract_all_kanji(self) -> set[str]:
        clean = ex_str.strip_html_and_bracket_markup(self._vocab.get_question() + self._vocab.forms.all_raw_string())
        return {char for char in clean if kana_utils.character_is_kanji(char)}

    @override
    def __repr__(self) -> str: return f"""main: {self.extract_main_form_kanji()}, all: {self.extract_all_kanji()}"""

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteKanji.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabNoteKanji
{
   readonly VocabNote _vocab;

   public VocabNoteKanji(VocabNote vocab) => _vocab = vocab;

   VocabNote Vocab => _vocab;

   public List<string> ExtractMainFormKanji()
   {
      var clean = StripHtmlAndBracketMarkup(Vocab.GetQuestion());
      return clean.Where(c => IsKanji(c)).Select(c => c.ToString()).ToList();
   }

   public HashSet<string> ExtractAllKanji()
   {
      var combined = Vocab.GetQuestion() + Vocab.Forms.AllRawString();
      var clean = StripHtmlAndBracketMarkup(combined);
      return clean.Where(c => IsKanji(c)).Select(c => c.ToString()).ToHashSet();
   }

   static string StripHtmlAndBracketMarkup(string text)
   {
      // Remove HTML tags
      var noHtml = Regex.Replace(text, "<[^>]*>", string.Empty);

      // Remove bracket markup like [kanji] or {brackets}
      var noBrackets = Regex.Replace(noHtml, @"\[[^\]]*\]", string.Empty);
      noBrackets = Regex.Replace(noBrackets, @"\{[^\}]*\}", string.Empty);

      return noBrackets;
   }

   static bool IsKanji(char c)
   {
      // Kanji Unicode ranges:
      // CJK Unified Ideographs: U+4E00 to U+9FFF
      // CJK Unified Ideographs Extension A: U+3400 to U+4DBF
      // CJK Compatibility Ideographs: U+F900 to U+FAFF
      var code = (int)c;
      return (code >= 0x4E00 && code <= 0x9FFF) ||
             (code >= 0x3400 && code <= 0x4DBF) ||
             (code >= 0xF900 && code <= 0xFAFF);
   }

   public override string ToString() => $"main: [{string.Join(", ", ExtractMainFormKanji())}], all: [{string.Join(", ", ExtractAllKanji())}]";
}
